File: src/main/java/net/liyze/basin/remote/RemoteServer.java
package net.liyze.basin.remote;

import net.liyze.basin.core.CommandParser;
import net.liyze.basin.core.Server;
import org.jetbrains.annotations.NotNull;
import org.smartboot.socket.MessageProcessor;
import org.smartboot.socket.extension.protocol.ByteArrayProtocol;
import org.smartboot.socket.transport.AioQuickServer;
import org.smartboot.socket.transport.WriteBuffer;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

import static net.liyze.basin.core.Main.LOGGER;

public class RemoteServer implements Server {
    public static final List<net.liyze.basin.core.Server> servers = new ArrayList<>();
    private final String token;
    private final int port;
    private final CommandParser REMOTE_Parser;
    public AioQuickServer server = null;

    /**
     * Init remote server at{@code <port>} with {@code <token>} and use {@code <remoteParser>} to parse remote command.
     */
    public RemoteServer(@NotNull String token, int port, CommandParser remoteParser) {
        servers.add(this);
        this.token = token;
        this.port = port;
        REMOTE_Parser = remoteParser;
    }

    /**
     * Stop remote Server
     */
    public void stop() {
        server.shutdown();
    }

    /**
     * Start remote server
     */
    public RemoteServer start() {
        MessageProcessor<byte[]> processor = (s, b) -> {
            Cipher cipher;
            String msg = "";
            WriteBuffer outputStream = s.writeBuffer();
            try {
                cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
                SecretKey keySpec = new SecretKeySpec(token.getBytes(StandardCharsets.UTF_8), "AES");
                cipher.init(Cipher.DECRYPT_MODE, keySpec);
                msg = new String(cipher.doFinal(b), StandardCharsets.UTF_8);
            } catch (Exception e) {
                try {
                    outputStream.write("Illegal BRC Request.".getBytes(StandardCharsets.UTF_8));
                } catch (IOException ex) {
                    LOGGER.error(ex.toString());
                }
                LOGGER.warn(e.toString());
            }
            if (!msg.startsWith("brc:")) {
                try {
                    byte[] bytes = "Illegal BRC Request.".getBytes(StandardCharsets.UTF_8);
                    LOGGER.warn("Illegal BRC Request: {} from {}", msg, s.getRemoteAddress().toString());
                    outputStream.writeInt(bytes.length);
                    outputStream.write(bytes);
                } catch (IOException e) {
                    LOGGER.warn("Illegal BRC Request: {}", msg);
                }
                return;
            }
            if (!REMOTE_Parser.sync().parse(msg.substring(4))) {
                try {
                    byte[] bytes = "Failed to run the command.".getBytes(StandardCharsets.UTF_8);
                    outputStream.writeInt(bytes.length);
                    outputStream.write(bytes);
                } catch (IOException ex) {
                    LOGGER.error(ex.toString());
                }
            } else {
                try {
                    byte[] bytes = "Okay".getBytes(StandardCharsets.UTF_8);
                    outputStream.writeInt(bytes.length);
                    outputStream.write(bytes);
                    LOGGER.info("BRC Request: {} from {}", msg, s.getRemoteAddress().toString());
                } catch (IOException e) {
                    LOGGER.warn(e.toString());
                }
            }

        };
        server = new AioQuickServer(port, new ByteArrayProtocol(), processor);
        server.setLowMemory(true);
        try {
            server.start();
        } catch (IOException e) {
            LOGGER.error("Failed to start server {}", e.toString());
        }
        return this;
    }
}

File: src/main/kotlin/net/liyze/basin/core/remote/RemoteServer.kt
/*
 * Copyright (c) 2023 Liyze09
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.liyze.basin.core.remote

import net.liyze.basin.core.CommandParser
import net.liyze.basin.core.Server
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.smartboot.socket.MessageProcessor
import org.smartboot.socket.extension.protocol.ByteArrayProtocol
import org.smartboot.socket.transport.AioQuickServer
import org.smartboot.socket.transport.AioSession
import java.io.IOException
import java.nio.charset.StandardCharsets
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.spec.SecretKeySpec

class RemoteServer(token: String, port: Int, remoteParser: CommandParser) : Server {
    private val token: String
    private val port: Int
    private val parser: CommandParser
    val server: AioQuickServer

    init {
        servers.add(this)
        this.token = token
        this.port = port
        parser = remoteParser
        val processor = MessageProcessor { s: AioSession, b: ByteArray ->
            val cipher: Cipher
            var msg = ""
            val outputStream = s.writeBuffer()
            try {
                cipher = Cipher.getInstance("AES/ECB/PKCS5Padding")
                val keySpec: SecretKey = SecretKeySpec(token.toByteArray(StandardCharsets.UTF_8), "AES")
                cipher.init(Cipher.DECRYPT_MODE, keySpec)
                msg = String(cipher.doFinal(b), StandardCharsets.UTF_8)
            } catch (e: Exception) {
                try {
                    outputStream.write("Illegal BRC Request.".toByteArray(StandardCharsets.UTF_8))
                } catch (ex: IOException) {
                    LOGGER.error(ex.toString())
                }
                LOGGER.warn(e.toString())
            }
            if (!msg.startsWith("brc:")) {
                try {
                    val bytes = "Illegal BRC Request.".toByteArray(StandardCharsets.UTF_8)
                    LOGGER.warn("Illegal BRC Request: {} from {}", msg, s.remoteAddress.toString())
                    outputStream.writeInt(bytes.size)
                    outputStream.write(bytes)
                } catch (e: IOException) {
                    LOGGER.warn("Illegal BRC Request: {}", msg)
                }
                return@MessageProcessor
            }
            if (!parser.sync().parseString(msg.substring(4))) {
                try {
                    val bytes = "Failed to run the command.".toByteArray(StandardCharsets.UTF_8)
                    outputStream.writeInt(bytes.size)
                    outputStream.write(bytes)
                } catch (ex: IOException) {
                    LOGGER.error(ex.toString())
                }
            } else {
                try {
                    val bytes = "Okay".toByteArray(StandardCharsets.UTF_8)
                    outputStream.writeInt(bytes.size)
                    outputStream.write(bytes)
                    LOGGER.info("BRC Request: {} from {}", msg, s.remoteAddress.toString())
                } catch (e: IOException) {
                    LOGGER.warn(e.toString())
                }
            }
        }
        server = AioQuickServer(port, ByteArrayProtocol(), processor)
    }

    /**
     * Stop remote Server
     */
    override fun stop() {
        server.shutdown()
    }

    /**
     * Start remote server
     */
    override fun start(): RemoteServer {

        try {
            server.start()
        } catch (e: IOException) {
            LOGGER.error("Failed to start server {}", e.toString())
        }
        return this
    }

    companion object {
        @JvmField
        val servers: MutableList<Server> = ArrayList()
        val LOGGER: Logger = LoggerFactory.getLogger(RemoteServer::class.java)
    }
}
